File: src/main/java/co/bitshifted/xapps/backstage/deploy/builders/MacDeploymentBuilder.java
/*
 * Copyright (c) 2020. Bitshift (http://bitshifted.co)
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package co.bitshifted.xapps.backstage.deploy.builders;

import co.bitshifted.xapps.backstage.deploy.DeploymentConfig;
import co.bitshifted.xapps.backstage.deploy.TargetDeploymentInfo;
import co.bitshifted.xapps.backstage.exception.DeploymentException;
import co.bitshifted.xapps.backstage.util.PackageUtil;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.io.FileUtils;
import org.slf4j.Logger;

import java.io.IOException;
import java.net.URISyntaxException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.nio.file.StandardOpenOption;

import static co.bitshifted.xapps.backstage.BackstageConstants.LAUNCHER_FILE_NAME_MAC;

/**
 * @author Vladimir Djurovic
 */
@Slf4j
public class MacDeploymentBuilder extends DeploymentBuilder {

	private static final String APP_BUNDLE_ARCHIVE_NAME = "app-bundle.zip";
	private static final String APP_BUNDLE_TEMPLATE = "/templates/" + APP_BUNDLE_ARCHIVE_NAME;
	private static final String APP_BUNDLE_JRE_DIR = "Contents/MacOS/jre";
	private static final String APP_BUNDLE_MACOS_DIR = "Contents/MacOS";
	private static final String APP_BUNDLE_RESOURCES_DIR = "Contents/Resources";

	public MacDeploymentBuilder(TargetDeploymentInfo targetDeploymentInfo) {
		super(targetDeploymentInfo);
	}

	@Override
	protected Path prepareDirectoryStructure() throws DeploymentException {
		try {
			var macBuildDir = macBuildDir();
			// create mac OS build directory
			Files.createDirectory(macBuildDir);
			// create skeleton for app bundle
			var templateUri = getClass().getResource(APP_BUNDLE_TEMPLATE).toURI();
			var appBundleArchive = macBuildDir.resolve(APP_BUNDLE_ARCHIVE_NAME);
			Files.copy(Path.of(templateUri), appBundleArchive, StandardCopyOption.REPLACE_EXISTING);
			var appBundlePath = PackageUtil.unpackZipArchive(appBundleArchive, config.macAppBundleName());
			log.debug("Extracted app bundle archive to {}", appBundlePath.toString());
			return appBundlePath;
		} catch(IOException| URISyntaxException ex) {
			throw new DeploymentException(ex);
		}

	}

	@Override
	protected void copyIcons(Path targetDir) {
		config.findMacIcons().stream().map(ic -> deploymentPackageDir.resolve(ic.getPath())).forEach(ic -> {
			try {
				FileUtils.copyToDirectory(ic.toFile(), targetDir.toFile());
			} catch(IOException ex) {
				log.error("Failed to copy icon {}", ic.toFile().getName());

			}
		});
	}

	@Override
	public void createDeployment() throws DeploymentException {
		try {
			var appBundlePath = prepareDirectoryStructure();
			// add modules for updates
			copySyncroModules();
			// create JRE image
			createJreImage(appBundlePath.resolve(APP_BUNDLE_JRE_DIR));
			// copy icon to resources
			copyIcons(appBundlePath.resolve(APP_BUNDLE_RESOURCES_DIR));
			copySplashScreen(appBundlePath.resolve(APP_BUNDLE_MACOS_DIR));
			// copy launcher
			copyLauncher(LAUNCHER_FILE_NAME_MAC, appBundlePath.resolve(APP_BUNDLE_MACOS_DIR));
			// create launcher configuration
			writeLauncherConfig(appBundlePath.resolve(APP_BUNDLE_MACOS_DIR));
			createInfoPlist(appBundlePath.resolve("Contents"), config);
			// create update package
			var updatePkgPath = createUpdatePackage(appBundlePath.resolve("Contents"));
			log.debug("Update package path: {}", updatePkgPath.toString());
			prepareForDownload(updatePkgPath);
		} catch(Exception ex) {
			log.error("Failed to create deployment", ex);
			throw new DeploymentException(ex);
		}

	}

	private Path macBuildDir() {
		return deploymentWorkDir.resolve("macos");
	}

	private void createInfoPlist(Path appBundleContentsPath, DeploymentConfig config) throws IOException {
		var infoPlist = appBundleContentsPath.resolve("Info.plist");
		var contents = Files.readString(infoPlist);
		var replaced = contents.replace("${app.name}", config.getAppName())
				.replace("${app.executable}", config.getExecutableFileName())
				.replace("${app.icon}", config.findMacIcons().get(0).getFileName())
				.replace("${app.id}", config.getAppId())
				.replace("${bundle.fqdn}", config.getAppId())
				.replace("${app.version}", config.getAppVersion());
		Files.writeString(infoPlist, replaced, StandardOpenOption.TRUNCATE_EXISTING);
	}

	@Override
	protected Path createUpdatePackage(Path bundleContentsDir) throws DeploymentException {
		try {
			var updateDir = macBuildDir().resolve("update");
			Files.createDirectories(updateDir);
			log.debug("Created update directory {}", updateDir.toString());
			var jreModules = bundleContentsDir.resolve("MacOS/jre/lib/modules");
			var moved = updateDir.resolve("modules");
			Files.move(jreModules, moved);
			var updatePath =  PackageUtil.packZipDeterministic(bundleContentsDir);
			// update package contents
			var updateContents = updateDir.resolve("contents.zip");
			Files.move(updatePath, updateContents);
			log.debug("Created contents update package at {}", updateContents.toString());
			// compress modules file
			var updateModules = PackageUtil.zipSingleFile(moved);
			log.debug("Created update modules file {}", updateModules.toString());
			Files.move(moved, jreModules);
			return updateDir;
		} catch(IOException ex) {
			throw new DeploymentException(ex);
		}

	}

	@Override
	protected Logger logger() {
		return log;
	}
}

File: src/main/kotlin/co/bitshifted/backstage/service/deployment/builders/MacDeploymentBuilder.kt
/*
 *
 *  * Copyright (c) 2022  Bitshift D.O.O (http://bitshifted.co)
 *  *
 *  * This Source Code Form is subject to the terms of the Mozilla Public
 *  * License, v. 2.0. If a copy of the MPL was not distributed with this
 *  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 */

package co.bitshifted.backstage.service.deployment.builders

import co.bitshifted.backstage.BackstageConstants
import co.bitshifted.backstage.exception.BackstageException
import co.bitshifted.backstage.exception.DeploymentException
import co.bitshifted.backstage.exception.ErrorInfo
import co.bitshifted.backstage.util.logger
import co.bitshifted.backstage.util.safeAppName
import co.bitshifted.ignite.common.model.OperatingSystem
import org.apache.commons.io.FileUtils
import java.io.FileWriter
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardCopyOption
import kotlin.io.path.absolutePathString
import kotlin.io.path.isDirectory
import kotlin.io.path.isExecutable
import kotlin.io.path.isSymbolicLink

class MacDeploymentBuilder(val builder: DeploymentBuilder){

    private val appBundleMacOSDirPath = "Contents/MacOS"
    private val appBundleResourcesDirPath = "Contents/Resources"
    private val infoPlistTemplate = "mac/info.plist.ftl"
    private val infoPlistFile = "Info.plist"
    private val createDmgTemplate = "mac/create-dmg.sh.ftl"
    private val createDmgScriptFileName = "create-dmg.sh"
    private val mbBytes = 1_000_000
    private val dsStoreInput = "/templates/mac/DS_Store"
    private val dsStoreFile = "DS_Store"
    private val backgroundImgInput = "/templates/mac/appforge-background.png"
    private val backgroundImageFile = "appforge-background.png"

    val logger = logger(this)
    lateinit var classpathDir: Path
    lateinit var modulesDir: Path
    lateinit var resourcesDir : Path
    lateinit var macOsDir : Path

    fun build() : Boolean {
        logger.info("Creating Mac OS X deployment in directory {}", builder.macDir)
        try {
            createDirectoryStructure()
            builder.copyDependencies(modulesDir, classpathDir, OperatingSystem.MAC)
            builder.copyResources(macOsDir)
            builder.buildJdkImage(macOsDir, modulesDir, OperatingSystem.MAC)
            copyLauncher()
            copyMacIcons()
            copySplashScreen()
            createInfoPlist()
            moveNonExecutableFiles()
            createInstaller()
            logger.info("Successfully created Mac OS X deployment in directory {}", builder.macDir)
            return true
        } catch(th : Throwable) {
            logger.error("Error building Mac OS X deployment", th)
            throw  th
        }
    }

    private fun createDirectoryStructure() {
        macOsDir = Files.createDirectories(builder.macDir.resolve(appBundleMacOSDirPath))
        classpathDir = Files.createDirectories(builder.macDir.resolve(appBundleMacOSDirPath).resolve(BackstageConstants.OUTPUT_CLASSPATH_DIR))
        logger.info("Created classpath directory at {}", classpathDir.toFile().absolutePath)
        modulesDir = Files.createDirectories(builder.macDir.resolve(appBundleMacOSDirPath).resolve(BackstageConstants.OUTPUT_MODULES_DIR))
        logger.info("Created modules directory at {}", modulesDir.toFile().absolutePath)
        resourcesDir = Files.createDirectories(builder.macDir.resolve(appBundleResourcesDirPath))
        logger.info("Created Resources directory in {}", resourcesDir.absolutePathString())
    }

    private fun copyLauncher() {
        val launcherPath = Path.of(builder.launchCodeDir.absolutePathString(),
            BackstageConstants.OUTPUT_LAUNCHER_DIST_DIR,
            BackstageConstants.LAUNCHER_NAME_MAC
        )
        logger.debug("Copying Mac OS X launcher from {} to {}", launcherPath.absolutePathString(), macOsDir.resolve(BackstageConstants.LAUNCHER_NAME_MAC))
        Files.copy(launcherPath, macOsDir.resolve(builder.builderConfig.deploymentConfig.applicationInfo.exeName), StandardCopyOption.COPY_ATTRIBUTES)
    }

    private fun copyMacIcons() {
        builder.builderConfig.deploymentConfig.applicationInfo.mac.icons.forEach {
            val name = if (it.target != null) it.target else it.source
            val target = resourcesDir.resolve(name)
            logger.debug("Icon target: {}", target.toFile().absolutePath)
            Files.createDirectories(target.parent)
            builder.builderConfig.contentService?.get(it.sha256 ?: throw BackstageException(ErrorInfo.EMPTY_CONTENT_CHECKSUM)).use {
                Files.copy(it, target, StandardCopyOption.REPLACE_EXISTING)
            }
        }
    }

    private fun copySplashScreen() {
        val splash = builder.builderConfig.deploymentConfig.applicationInfo.splashScreen
        if (splash!= null) {
            val name = if (splash.target != null) splash.target else splash.source
            val target = resourcesDir.resolve(name)
            logger.debug("Splash screen target: {}", target.toFile().absolutePath)
            Files.createDirectories(target.parent)
            builder.builderConfig.contentService?.get(splash.sha256 ?: throw BackstageException(ErrorInfo.EMPTY_CONTENT_CHECKSUM)).use {
                Files.copy(it, target, StandardCopyOption.REPLACE_EXISTING)
            }
            // create symlink to splash screen
            Files.createSymbolicLink(macOsDir.resolve(name), macOsDir.relativize(resourcesDir.resolve(target.fileName)))
        }
    }

    private fun getTemplateData() : MutableMap<String, Any> {
        val data = mutableMapOf<String, Any>()
        data["appName"] = builder.builderConfig.deploymentConfig.applicationInfo.name
        data["appExecutable"] = builder.builderConfig.deploymentConfig.applicationInfo.exeName
        data["appIcon"] = builder.builderConfig.deploymentConfig.applicationInfo.mac.icons[0].target
        data["appId"] = builder.builderConfig.deploymentConfig.applicationId
        data["bundleFqdn"] = builder.builderConfig.deploymentConfig.applicationId
        data["appVersion"] = builder.builderConfig.deploymentConfig.version
        data["installerDir"] = builder.installerDir.absolutePathString()
        data["appSafeName"] = safeAppName(builder.builderConfig.deploymentConfig.applicationInfo.name)
        data["macOutputDir"] = builder.macDir.absolutePathString()
        val dirSizeBytes = FileUtils.sizeOfDirectory(builder.macDir.toFile())
        data["sizeInMb"] = (dirSizeBytes / mbBytes) + 2
        return data
    }

    private fun createInfoPlist() {
        val data = getTemplateData()
        val template = builder.freemarkerConfig.getTemplate(infoPlistTemplate)
        val targetPath = macOsDir.parent.resolve(infoPlistFile)
        val writer = FileWriter(targetPath.toFile())
        writer.use {
            template.process(data, writer)
        }
    }

    // Mac OS bundle signing fails if there are non-executable files in Content/MacOS directory.
    // this function moves them
    private fun moveNonExecutableFiles() {
        Files.list(macOsDir).forEach {
            if (it.isDirectory()) {
                try {
                    FileUtils.moveDirectory(it.toFile(), resourcesDir.resolve(it.fileName).toFile())
                    createSymlink(macOsDir.resolve(it.fileName), macOsDir.relativize(resourcesDir.resolve(it.fileName)))
                } catch(ex : IOException) {
                    logger.error("Failed to move directory", ex)
                }
            } else if (!it.isExecutable() && !it.isSymbolicLink()) {
                Files.move(it, resourcesDir.resolve(it.fileName))
                Files.createSymbolicLink(macOsDir.resolve(it.fileName), macOsDir.relativize(resourcesDir.resolve(it.fileName)))
            }

        }
    }

    private fun createSymlink(source : Path, target : Path) {
        logger.debug("Creating symlink: {} -> {}", source.absolutePathString(), target.toString())
        Files.createSymbolicLink(source, target)
    }

    private fun createInstaller() {
        logger.info("Creating Mac OS X installer")
        val dsStore = this.javaClass.getResourceAsStream(dsStoreInput)
        Files.copy(dsStore, builder.installerDir.resolve(dsStoreFile))
        val bgImage = this.javaClass.getResourceAsStream(backgroundImgInput)
        Files.copy(bgImage, builder.installerDir.resolve(backgroundImageFile))
        val data = getTemplateData()
        val template = builder.freemarkerConfig.getTemplate(createDmgTemplate)
        val installerFile = builder.installerDir.resolve(createDmgScriptFileName)
        val writer = FileWriter(installerFile.toFile())
        writer.use {
            template.process(data, writer)
        }
        installerFile.toFile().setExecutable(true)
        // invoke installer creation
        val pb = ProcessBuilder("./$createDmgScriptFileName")
        pb.directory(builder.installerDir.toFile())
        pb.environment().put("PWD", builder.installerDir.absolutePathString())
        logger.debug("create-dmg.sh: working directory={}", builder.installerDir.absolutePathString())
        val process = pb.start()
        if (process.waitFor() == 0) {
            logger.info(process.inputReader().use { it.readText() })
            logger.info("Mac OS DMG image created successfully")
        } else {
            logger.error("Error encountered while creating Mac OS DMG image. Details:")
            logger.error(process.inputReader().use { it.readText() })
            logger.error(process.errorReader().use { it.readText() })
            throw DeploymentException("Failed to build Mac DMG image")
        }
    }
}